starlite/controller/generic_controller.py

```diff
@@ -0,0 +1,70 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Generic, TypeVar
+
+from typing_extensions import Self, get_args, get_origin
+
+from .base import Controller
+
+if TYPE_CHECKING:
+    from typing import Any
+
+
+__all__ = ("GenericController",)
+
+T = TypeVar("T")
+
+
+class GenericController(Controller, Generic[T]):
+    """Controller type that supports generic inheritance hierarchies."""
+
+    _data_type: type[T]
+
+    def __class_getitem__(cls, data_type: type[T]) -> type[Self]:
+        return type(f"GenericController[{data_type.__name__}]", (cls,), {"_data_type": data_type})
+
+    def _get_new_args(self, args: tuple[Any, ...], origin: Any | None) -> tuple[Any, ...] | None:
+        if not args and origin:
+            return None
+
+        type_var_found = False
+        new_args = []
+        for arg in args:
+            if isinstance(arg, TypeVar):
+                new_args.append(self._data_type)
+                type_var_found = True
+                continue
+            new_args.append(arg)
+
+        if not type_var_found:
+            return None
+
+        return tuple(new_args)
+
+    @staticmethod
+    def _rebuild_annotation(original_annotation: Any, origin: Any | None, args_tuple: tuple[Any, ...]) -> Any:
+        if origin is None:
+            raise RuntimeError("Unexpected origin value")
```

Comment:
Why not ImproperlyConfigured?

Reply:
I don't think it should ever hold true - but I am not confident enough to say that a user has improperly configured something if it does. 

I want to narrow the type of origin from optional to not-optional, and explicitly find out when my assumption that it should never be `None` if untrue. So it is an error that I want to be an internal error, but not necessarily a user-did-something-wrong error.

I'll just change it to StarliteException.